src/modalities/models/coca/multi_modal_decoder.py

```diff
@@ -0,0 +1,90 @@
+from typing import Dict
+
+import torch
+import xformers.ops as xops
+from torch import nn
+
+from modalities.models.gpt2.gpt2_model import ActivationType, AttentionConfig, LayerNorm, TransformerMLP
+from modalities.models.model import NNModel
+from modalities.nn.attention import Attention
+
+
+class MultiModalBlock(nn.Module):
+    def __init__(
+        self,
+        n_embd: int,
+        bias: bool,
+        epsilon: float,
+        activation: ActivationType,
+        n_head: int,
+        dropout: float,
+        ffn_hidden: int,
+    ):
+        super().__init__()
+        self.ln_1 = LayerNorm(ndim=n_embd, bias=bias, epsilon=epsilon)
+        self.attn = Attention(n_embd, n_head, bias, is_causal=False, use_cross_attention=True)
+        self.ln_2 = LayerNorm(ndim=n_embd, bias=bias, epsilon=epsilon)

```

Comment:
Don't we need a self-attention layer with causal masking first, before the cross-attention layer?

Reply:
For [open CoCa](https://github.com/mlfoundations/open_clip/blob/main/src/open_clip/transformer.py#L188) they also only used Cross-Attention. However, an additional self-attention layer will probably help with loading pretrained LLM.

Reply:
Had a brief look, I think they actually also use the self-attention layer (called `resblock`), see [here](https://github.com/mlfoundations/open_clip/blob/73fa7f03a33da53653f61841eb6d69aef161e521/src/open_clip/transformer.py#L790)